packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "PerformanceObserver.h"
+#include "PerformanceObserverRegistry.h"
+#include "PerformanceEntryReporter.h"
+
+namespace facebook::react {
+
+PerformanceObserver::~PerformanceObserver() {
+  if (auto registry = registry_.lock()) {
+    registry->removeObserver(*this);
+  }
+}
+
+void PerformanceObserver::pushEntry(const PerformanceEntry& entry) {
+  buffer_.add(entry);
+}
+
+std::vector<PerformanceEntry> PerformanceObserver::takeRecords() {
+  return buffer_.consume();
+}
+
+bool PerformanceObserver::isObserving(PerformanceEntryType type) const {
+  return observedTypes_.contains(type);
+}
+
+void PerformanceObserver::observe(PerformanceEntryType type, bool buffered) {
+  // we assume that `type` was checked on JS side and is correct
+  observedTypes_.clear();
+  observedTypes_.insert(type);
+
+  requiresDroppedEntries_ = true;
+
+  if (buffered) {
+    auto& reporter = PerformanceEntryReporter::getInstance();
+    reporter->getBuffer(type).getEntries(std::nullopt, buffer_.getEntries());
+    scheduleFlushBuffer();
+  }
+}
+
+void PerformanceObserver::observe(std::unordered_set<PerformanceEntryType> types) {
+  observedTypes_ = std::move(types);
+  requiresDroppedEntries_ = true;
+}
+
+bool PerformanceObserver::shouldAdd(const PerformanceEntry& entry) const {
+  return buffer_.shouldAdd(entry);
```

Comment:
We're not checking if the observer is actually observing the type of entry we're receiving, which is incorrect. We should check 2 things here:
1. That the observer is observing the type of the entry
2. If the entry is of type event, if the entry has a duration higher than the threshold specified by the observer.

The buffer for observers is unlimited so we shouldn't be checking anything specific about it (also simplifies making it just a simple `std::vector`, which we can swap in `takeRecords`)